cron/appstats.go

```diff
@@ -22,33 +13,34 @@ type App struct {
 
 
 
-func GetTopApps(db *sql.DB, amount int, days int) ([]App, error){
+func GetTopApps(db *sql.DB, amount int, startDate string, endDate string) ([]App, error){
 	var name *string
 	var appID string
 	var appCount int
 
 	query := `SELECT app_name, app_id, count(*) AS job_count FROM jobs
-           WHERE start_date >= (now() - ($2 || ' DAY')::INTERVAL )
+           WHERE start_date >= ($2 :: DATE)
+           AND start_date <= ($3 :: DATE)
            AND app_id != '1e8f719b-0452-4d39-a2f3-8714793ee3e6'
            GROUP BY app_name, app_id
            ORDER BY job_count DESC
            LIMIT $1`
-	rows, err := db.Query(query, amount, days)
+	rows, err := db.Query(query, amount, startDate, endDate)
 
 	if err != nil {
 		return nil, err
 	}
 
 	defer rows.Close()
 
-	apps := make([]App, amount)
+	var apps []App
 	for i := 0; rows.Next(); i++{
 		err := rows.Scan(&name, &appID, &appCount)
 		if err != nil {
 			return nil, err
 		}
 
-		apps[i] = App{getStringValue(name), appID, appCount}
+		apps = append(apps, App{getStringValue(name), appID, appCount})
 		output := fmt.Sprintf("App name %[1]v App ID %[2]v App Count %[3]v", getStringValue(name), appID, appCount)
 		fmt.Println(output)
```

Comment:
Wait printing to console isn't a good way to log data? Well school has been a huge waste of money. Haha just kidding would the timestamps be part of the json? And where would these logs be sent to. Probably the lowest priority but good to keep in mind.

Reply:
So in fact I'm not even telling you not to print to console, :laughing:, just to add something that makes it structured. We have a system (right now, only in prod -- it probably needs to be revisited, especially for QA, where it went away when we re-did our k8s deployment) that looks at the output of containers and sends it to Elasticsearch. Then, theoretically, it can be accessed in Kibana, filtered, etc.

The easiest way to do it would be:

1. add `"github.com/sirupsen/logrus"` to the top-level import
2. Add something like this in the main namespace:

```
var log = logrus.WithFields(logrus.Fields{
    "service": "de-stats",
})

func init() {
    logrus.SetFormatter(&logrus.JSONFormatter{})
}
```

Then, you can use methods on that new `log` object to create nice logs, e.g.: `log.Fatal(err.Error())` logs and then quits with an error (only do that directly in `main`, though), or `log.Infof("Something or other with %s placeholders", someVariable)` or [all manner of other stuff](https://godoc.org/github.com/sirupsen/logrus#Entry). (Technically the `service` thing isn't even that needed, you could just use `logrus.Infof` and such, but it's nice to have the service name in the log line, and we do that in our other stuff.). Note in that list that you also get logging levels from this -- it means you can add a whole bunch of `log.Debug`/`log.Debugf` lines that won't print out by default, and add a command-line argument that turns them on. You'll probably need to add command-line arguments to this service eventually, in order to support connecting to different DBs in prod vs. QA. But it's just done nicely in the same init function -- see [infosquito2's version](https://github.com/cyverse-de/infosquito2/blob/master/main.go#L64-L72) for an example of how that looks.

Since you're doing things across several namespaces, you'll probably want to pass that log object around similarly to the DB as we're discussing in other comments. I don't remember if you'll need to reproduce that `init()` func in each namespace or not, but that's easily enough tested (the `log` variable should probably not be reproduced). (I hope not, since it's nice to only have to do stuff one place, but dunno).

One other nice thing you can do is, if you want, add extra fields to the logs within some context. For example, in [infosquito2](https://github.com/cyverse-de/infosquito2/blob/master/reindex.go#L282-L284) we partition things into a set of prefixes and then process each prefix separately, so in the things processing a given prefix, we add that prefix to every log line too so it's easier to figure out what's going on if the error has to do with a specific subset of the data.

We may at some point want to move away from JSON-formatted logs (we've talked about it), but using a proper logging tool also makes this easier. For example, JSON logging was never set up for the `user-info` service, so it logs things like: `time="2019-12-06T12:07:01-07:00" level=info msg="Getting user preferences for ipctest@iplantcollaborative.org" service=preferences` instead of formatting that as JSON. Making that change is just about changing or removing that `init` function above, so it's much easier to change things around.

Hopefully you're convinced it's worth the changeover, anyway. We can certainly help you make it happen if you're not sure from my wall of text above.